src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -1346,53 +1370,51 @@ static void print_signal_set_short(outputStream* st, const sigset_t* set) {
   st->print("%s", buf);
 }
 
-void PosixSignals::print_signal_handler(outputStream* st, int sig,
-                                 char* buf, size_t buflen) {
-  struct sigaction sa;
-  sigaction(sig, NULL, &sa);
-
-  // See comment for SA_RESTORER_FLAG_MASK
-  LINUX_ONLY(sa.sa_flags &= SA_RESTORER_FLAG_MASK;)
+static void print_single_signal_handler(outputStream* st, int sig,
+                                        const struct sigaction* act,
+                                        char* buf, size_t buflen) {
 
   st->print("%10s: ", os::exception_name(sig, buf, buflen));
 
-  address handler = get_signal_handler(&sa);
-
-  if (handler == CAST_FROM_FN_PTR(address, SIG_DFL)) {
+  address handler = get_signal_handler(act);
+  if (HANDLER_IS_DFL(handler)) {
     st->print("SIG_DFL");
-  } else if (handler == CAST_FROM_FN_PTR(address, SIG_IGN)) {
+  } else if (HANDLER_IS_IGN(handler)) {
     st->print("SIG_IGN");
   } else {
-    print_signal_handler_name(st, handler, buf, O_BUFLEN);
+    print_signal_handler_name(st, handler, buf, buflen);
   }
 
-  st->print(", sa_mask[0]=");
-  print_signal_set_short(st, &sa.sa_mask);
+  st->print(", mask=");
+  print_signal_set_short(st, &(act->sa_mask));
 
-  address rh = VMError::get_resetted_sighandler(sig);
-  // May be, handler was resetted by VMError?
-  if (rh != NULL) {
-    handler = rh;
-    // See comment for SA_RESTORER_FLAG_MASK
-    sa.sa_flags = VMError::get_resetted_sigflags(sig) LINUX_ONLY(& SA_RESTORER_FLAG_MASK);
-  }
+  st->print(", flags=");
+  int flags = act->sa_flags;
+  // On Linux, hide the SA_RESTORE flag
+  LINUX_ONLY(flags &= SA_RESTORER_FLAG_MASK;)
+  print_sa_flags(st, flags);
 
-  // Print textual representation of sa_flags.
-  st->print(", sa_flags=");
-  print_sa_flags(st, sa.sa_flags);
+}
 
-  // Check: is it our handler?
-  if (handler == CAST_FROM_FN_PTR(address, (sa_sigaction_t)javaSignalHandler) ||
-      handler == CAST_FROM_FN_PTR(address, (sa_sigaction_t)SR_handler)) {
-    // It is our signal handler
-    // check for flags, reset system-used one!
-    if ((int)sa.sa_flags != get_our_sigflags(sig)) {
-      st->print(
-                ", flags was changed from " PTR32_FORMAT ", consider using jsig library",
-                get_our_sigflags(sig));
```

Comment:
It is not clear to me that this check still remains somewhere.

Reply:
I removed this check since I did not think it very useful. We have the checking code with Xcheck:jni, which does the same check. 

But I am now reconsidering. If we run without Xcheck, it still is useful. I'll re-add this test.